src/util.c

```diff
@@ -967,3 +968,54 @@ int utilTest(int argc, char **argv, int accurate) {
     return 0;
 }
 #endif
+
+int createPipe(int fds[2], int read_flags, int write_flags) {
+#if defined(__linux__) || defined(__FreeBSD__) 
+    /* When it leverages the pipe2() to create pipe, 
+     * there is no harm to set O_CLOEXEC to prevent fd leaks. */
+    int flags = O_CLOEXEC;
+
+    if ((read_flags & O_NONBLOCK) && (write_flags & O_NONBLOCK))
+        flags |= O_NONBLOCK;
+
+    int ret = pipe2(fds, flags);
+    if (ret && errno != ENOSYS) {
+        return -1;
+    } else if (ret) {
+        if (pipe(fds))
+            return -1;
+        if (read_flags & O_CLOEXEC)
+            if ((anetCloexec(fds[0]) != ANET_OK))
+                goto error;
+        if (write_flags & O_CLOEXEC)
+            if ((anetCloexec(fds[1]) != ANET_OK))
+                goto error;
+        }
+
+    if (!ret && (flags & O_NONBLOCK))
+        return 0;
+#else
+    if (pipe(fds))
+        return -1;
+    if (read_flags & O_CLOEXEC)
+        if ((anetCloexec(fds[0]) != ANET_OK))
+            goto error;
+    if (write_flags & O_CLOEXEC)
+        if ((anetCloexec(fds[1]) != ANET_OK))
+            goto error;
+#endif
+
+    if (read_flags & O_NONBLOCK)
+        if ((anetNonBlock(NULL, fds[0]) != ANET_OK))
+            goto error;
+    if (write_flags & O_NONBLOCK)
+        if ((anetNonBlock(NULL, fds[1]) != ANET_OK))
+            goto error;
+
+    return 0;
```

Comment:
> This shouldn't be done here if pipe2() has already set O_NONBLOCK. It's extra syscalls which you wanted to avoid using pipe2, right?

I think you miss the above code:
```c
    if (!ret && (flags & O_NONBLOCK))
        return 0;
```

Reply:
But I can try to simplify the code further.

Reply:
> ... and we can easily allow any other flags as well (all except CLOEXEC can be set using F_SETFL) just by changing the last part like this:

I don't think this is necessary, we should only focus on `O_NONBLOCK` and `O_CLOEXEC` which are currently used.

Reply:
![image](https://user-images.githubusercontent.com/7496278/133914793-d186723e-01b7-4cfd-acc2-7ea9cb6ea088.png)
We need to resolve the compliation of "unused labels" in redis CI.

Reply:
Oh, just put the pipe label before `#endif`

Reply:
For 'common' I think it's not to hard to solve...

Reply:
what about label "common"?

Reply:
It seems to me that the lable "common" will fail the build on mac anyway.

Reply:
The code now looks like this:

![image](https://user-images.githubusercontent.com/7496278/133922215-50691045-2492-41e5-ab65-63388e230828.png)